SmartDeviceLink/public/SDLVideoStreamingCapability.m

```diff
@@ -90,6 +91,113 @@ - (void)setScale:(nullable NSNumber<SDLFloat> *)scale {
+    return [[SDLImageResolution alloc] initWithWidth:(uint16_t)size.width height:(uint16_t)size.height];
+}
+
+- (NSArray<SDLImageResolution *> *)allImageResolutions {
+    NSMutableArray<SDLImageResolution *> *resolutions = [NSMutableArray arrayWithCapacity:self.additionalVideoStreamingCapabilities.count + 1];
+    if (self.preferredResolution) {
+        [resolutions addObject:self.preferredResolution];
+    }
+    for (SDLVideoStreamingCapability *nextCapability in self.additionalVideoStreamingCapabilities) {
+        if (nextCapability.preferredResolution) {
+            [resolutions addObject:nextCapability.preferredResolution];
+        }
+    }
+    return resolutions;
+}
+
+- (NSArray<SDLImageResolution *> *)allImageResolutionsScaled {
+    NSMutableArray<SDLImageResolution *> *resolutions = [NSMutableArray arrayWithCapacity:self.additionalVideoStreamingCapabilities.count + 1];
+    SDLImageResolution *imgResolution = [self makeImageResolution];
+    if (imgResolution) {
+        [resolutions addObject:imgResolution];
+    }
+    for (SDLVideoStreamingCapability *nextCapability in self.additionalVideoStreamingCapabilities) {
+        SDLImageResolution *imgResolution = [nextCapability makeImageResolution];
+        if (imgResolution) {
+            [resolutions addObject:imgResolution];
+        }
+    }
+    return resolutions;
+}
+
+#define EQUAL_RES ((nil == self.preferredResolution && nil == cap2.preferredResolution) ? YES : [self.preferredResolution isEqual:cap2.preferredResolution])
+#define EQUAL_NUM(property) ((nil == self.property && nil == cap2.property) ? YES : [self.property isEqualToNumber:cap2.property])
+
+- (BOOL)isEqual:(id)object {
+    if (![object isKindOfClass:self.class]) {
+        return NO;
+    }
+    typeof(self) cap2 = object;
+    if (!EQUAL_RES) {
+        return NO;
+    }
+    if (!EQUAL_NUM(maxBitrate)) {
+        return NO;
+    }
+    if (!EQUAL_NUM(diagonalScreenSize)) {
+        return NO;
+    }
+    if (!EQUAL_NUM(pixelPerInch)) {
+        return NO;
+    }
+    if (!EQUAL_NUM(scale)) {
+        return NO;
+    }
+    return YES;
+}
+
+#undef EQUAL_RES
+#undef EQUAL_NUM
+
+- (NSString *)description {
+    NSMutableString *formats = [NSMutableString string];
+    for (SDLVideoStreamingFormat * f in self.supportedFormats) {
+        [formats appendFormat:@"%@; ", f];
+    }
+    return [NSString stringWithFormat:@"<%@:%p>{\n\tsupportedFormats:%@\n\tpreferredResolution:%@\n\tmaxBitrate:%@\n\thapticSpatialDataSupported:%@\n\tdiagonalScreenSize:%@\n\tpixelPerInch:%@\n\tscale:%@\n\tadditionalVideoStreamingCapabilities[%d]:%@ }", NSStringFromClass(self.class), self, formats, self.preferredResolution, self.maxBitrate, self.hapticSpatialDataSupported.boolValue ? @"YES" : @"NO", self.diagonalScreenSize, self.pixelPerInch, self.scale, (int)self.additionalVideoStreamingCapabilities.count, self.additionalVideoStreamingCapabilities];
```

Comment:
is there a way you can break each component of the formatted string into stand-alone string variables and then concat them?

Reply:
yes sure there is